File: lib/ExpediaBookings/src/main/java/com/expedia/bookings/utils/NumberUtils.java
package com.expedia.bookings.utils;

import java.math.BigDecimal;
import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

public class NumberUtils {

	public static Double parseDoubleSafe(String str) {
		if (Strings.isEmpty(str)) {
			return null;
		}

		try {
			return Double.parseDouble(str);
		}
		catch (NumberFormatException e) {
			return null;
		}
	}

	public static int getPercentagePaidWithPointsForOmniture(BigDecimal fraction, BigDecimal total) {
		if (total.equals(BigDecimal.ZERO)) {
			throw new IllegalArgumentException("Total cannot be zero while calculating percentage");
		}
		return fraction.divide(total, 2, BigDecimal.ROUND_HALF_UP).multiply(new BigDecimal("100")).intValue();
	}

	public static List<Integer> getRandomNumberList(int size) {
		List<Integer> numberList = new ArrayList<>();
		for (int index = 0; index < size; index++) {
			numberList.add(index);
		}
		Collections.shuffle(numberList);
		return numberList;
	}

	public static float round(float number, int precision) {
		float shiftNumberByPlaces = (float) Math.pow(10, precision);
		return Math.round(number * shiftNumberByPlaces) / shiftNumberByPlaces;
	}
}

File: lib/ExpediaBookings/src/main/java/com/expedia/bookings/utils/NumberUtils.kt
package com.expedia.bookings.utils

import java.math.BigDecimal
import java.text.NumberFormat
import java.util.Locale

object NumberUtils {

    private var localeToNumberFormatMap = HashMap<Int, NumberFormat>()

    @JvmStatic
    fun parseDoubleSafe(str: String): Double? {
        if (Strings.isEmpty(str)) {
            return null
        }

        try {
            return java.lang.Double.parseDouble(str)
        } catch (e: NumberFormatException) {
            return null
        }
    }

    @JvmStatic
    fun getPercentagePaidWithPointsForOmniture(fraction: BigDecimal, total: BigDecimal): Int {
        if (total == BigDecimal.ZERO) {
            throw IllegalArgumentException("Total cannot be zero while calculating percentage")
        }
        return fraction.divide(total, 2, BigDecimal.ROUND_HALF_UP).multiply(BigDecimal("100")).toInt()
    }

    @JvmStatic
    fun round(number: Float, precision: Int): Float {
        val shiftNumberByPlaces = Math.pow(10.0, precision.toDouble()).toFloat()
        return Math.round(number * shiftNumberByPlaces) / shiftNumberByPlaces
    }

    @JvmStatic
    fun localeBasedFormattedNumber(number: Number): String {
        val locale = Locale.getDefault()
        val key = locale.hashCode()
        var nf = localeToNumberFormatMap.get(key)
        if (nf == null) {
            nf = NumberFormat.getInstance(locale) as NumberFormat
            localeToNumberFormatMap.put(key, nf)
        }
        return nf.format(number)
    }
}
